vanilla/src/launch/java/org/spongepowered/vanilla/launch/plugin/resolver/DependencyResolver.java

```diff
@@ -0,0 +1,228 @@
+/*
+ * This file is part of Sponge, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.vanilla.launch.plugin.resolver;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.spongepowered.api.util.Tuple;
+import org.spongepowered.plugin.PluginCandidate;
+import org.spongepowered.plugin.PluginResource;
+import org.spongepowered.plugin.metadata.PluginDependency;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public final class DependencyResolver {
+
+    public static <T extends PluginResource> ResolutionResult<T> resolveAndSortCandidates(final Collection<PluginCandidate<T>> candidates) {
+        final Map<String, Node<T>> nodes = new HashMap<>();
+        final ResolutionResult<T> resolutionResult = new ResolutionResult<>();
+        for (final PluginCandidate<T> candidate : candidates) {
+            final String id = candidate.metadata().id();
+            // If we already have an entry, this is now a duplicate ID situation.
+            if (nodes.containsKey(id)) {
```

Comment:
`containsKey` + `put` -> `putIfAbsent` ?

Reply:
Don't want to create a new `Node` if we don't have to.